agent/xds/listeners_ingress.go

```diff
@@ -154,21 +137,107 @@ func (s *ResourceGenerator) makeIngressGatewayListeners(address string, cfgSnap
 	return resources, nil
 }
 
-func resolveListenerSDSConfig(cfgSnap *proxycfg.ConfigSnapshot, listenerCfg structs.IngressListener) (*structs.GatewayTLSSDSConfig, error) {
+func makeDownstreamTLSContextFromSnapshotListenerConfig(cfgSnap *proxycfg.ConfigSnapshot, listenerCfg structs.IngressListener) (*envoy_tls_v3.DownstreamTlsContext, error) {
+	var downstreamContext *envoy_tls_v3.DownstreamTlsContext
+
+	tlsContext, err := makeCommonTLSContextFromSnapshotListenerConfig(cfgSnap, listenerCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	if tlsContext != nil {
+		downstreamContext = &envoy_tls_v3.DownstreamTlsContext{
+			CommonTlsContext:         tlsContext,
+			RequireClientCertificate: &wrappers.BoolValue{Value: false},
+		}
+	}
+
+	return downstreamContext, nil
+}
+
+func makeCommonTLSContextFromSnapshotListenerConfig(cfgSnap *proxycfg.ConfigSnapshot, listenerCfg structs.IngressListener) (*envoy_tls_v3.CommonTlsContext, error) {
+	var tlsContext *envoy_tls_v3.CommonTlsContext
+
+	// Enable connect TLS if it is enabled at the Gateway or specific listener
+	// level.
+	gatewayTLSCfg := cfgSnap.IngressGateway.TLSConfig
+
+	// TODO: should it be possible to explicitly _disable_ TLS on a listener if it's
+	// enabled on the gateway?
+	connectTLSEnabled := gatewayTLSCfg.Enabled ||
+		(listenerCfg.TLS != nil && listenerCfg.TLS.Enabled)
+
+	tlsCfg, err := resolveListenerTLSConfig(&gatewayTLSCfg, listenerCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	if tlsCfg.SDS != nil {
+		// Set up listener TLS from SDS
+		tlsContext = makeCommonTLSContextFromGatewayTLSConfig(*tlsCfg)
+	} else if connectTLSEnabled {
+		tlsContext = makeCommonTLSContextFromLeaf(cfgSnap, cfgSnap.Leaf(), makeTLSParametersFromGatewayTLSConfig(*tlsCfg))
+	}
+
+	return tlsContext, nil
+}
+
+// func resolveGatewayServiceTLSConfig(mergedListenerTLSConfig *structs.GatewayTLSConfig, gatewayServiceCfg structs.GatewayServiceTLSConfig) (*structs.GatewayServiceTLSConfig, error) {
+// 	return nil, nil
+// }
```

Comment:
stray comment?

Reply:
Related to question at https://github.com/hashicorp/consul/pull/11576#discussion_r768107890